src/argparse.lua

```diff
@@ -1087,6 +1087,230 @@ function Parser:get_help()
    return table.concat(blocks, "\n\n")
 end
 
+function Parser:add_complete(value)
+   if value then
+      assert(type(value) == "string" or type(value) == "table",
+         ("bad argument #1 to 'add_complete' (string or table expected, got %s)"):format(type(value)))
+   end
+
+   local complete = self:option()
+      :description "Output a shell completion script for the specified shell."
+      :args(1)
+      :choices {"bash", "zsh", "fish"}
+      :action(function(_, _, shell)
+         print(self["get_" .. shell .. "_complete"](self))
+         os.exit(0)
+      end)
+
+   if value then
+      complete = complete(value)
+   end
+
+   if not complete._name then
+      complete "--completion"
+   end
+
+   return self
+end
+
+local function get_short_description(element)
+   local short = element._description:match("^(.-)%.%s")
+   return short or element._description:match("^(.-)%.?$")
+end
+
+local function get_options(parser)
+   local options = {}
+   for _, option in ipairs(parser._options) do
+      for _, alias in ipairs(option._aliases) do
+         table.insert(options, alias)
+      end
+   end
+   return table.concat(options, " ")
+end
+
+local function get_commands(parser)
+   local commands = {}
+   for _, command in ipairs(parser._commands) do
+      for _, alias in ipairs(command._aliases) do
+         table.insert(commands, alias)
+      end
+   end
+   return table.concat(commands, " ")
+end
+
+function Parser:_bash_option_args(buf, indent)
+   local opts = {}
+   for _, option in ipairs(self._options) do
+      if option._choices or option._minargs > 0 then
+         local pattern = ("%s)"):format(table.concat(option._aliases, "|"))
+         local compreply
+         if option._choices then
+            compreply = ('COMPREPLY=($(compgen -W "%s" -- "$cur"))')
```

Comment:
I think you might need to wrap compgen invocations in `IFS` resets.
I'm wondering if using a separator of e.g. a null byte might be the best choice?

Reply:
Bash doesn't allow a null byte as it would terminate the string. Also it seems there's no simple way to allow completions with whitespace, so a space should be fine.

Reply:
This works for me locally with bash:
```sh
IFS=$'\0' compgen -W $'foo\0bar' -- "fo"
```

Reply:
Nevermind, it doesn't do what I thought.

Reply:
What do you think of still doing IFS resets (to ' ')?

Reply:
Turns out a newline is needed because compgen prints completions one per line.